Addressing `comet/comet#1174`. First version (#14)
cometbft/knowledge-base: protocols/abci/addressing-1174.md

# Allowing Non-Determinism in `ProcessProposal` (a.k.a. addressing #1174)

## Context

### Consensus Properties

Byzantine Fault Tolerant (BFT) Consensus is usually specified by the following invariants:

- _agreement_: no two correct processes decide differently.
- _validity_: the decided block `v` satisfies the predefined predicate `valid(v)`.
- _termination_: all correct processes eventually decide.

The consensus algorithm implemented in CometBFT (Tendermint) guarantees these invariants for each height.

> 

Author: Thanks Josef for the comment. That matches my understanding, and is part of what we are requiring (one of the determinism requirements) in the ABCI++ spec.

I'll include this text annotation in the write-up, and see how I can integrate it

### ABCI interface

In the version of ABCI (`v0.17.0`) that existed before ABCI 1.0 and 2.0 (a.k.a. ABCI++),
the implementation of the `valid(v)` predicate was totally internal to CometBFT.
The application had no direct say on the validity of a block,
although it could (and still can) influence it indirectly via the best-effort ABCI call `CheckTx`.

With the evolution of ABCI to ABCI 1.0 and 2.0, the `valid(v)` predicate now has two components:

- the validity checks performed directly by CometBFT on blocks
  (block format, hashes, etc; the same as in ABCI `v0.17.0`)
- the validity checks that now the application can perform as part of `ProcessProposal`

With these new validity checks:

- consensus _agreement_ is not affected since the new structure of `valid(v)` in ABCI 1.0
  does not accept any block `v` that would have been rejected in ABCI 0.17.0.

>   The new structure of `valid(v)` is thus more restrictive for any block `v`.

Author: @lasarojc your proposal seems good to me. However, one thing I wanted to include in the original text is giving a hint on _why_ agreement is not affected.
If we don't accept any proposal that was previously rejected, we can't be allowing runs violating agreement that were previously prevented.
I'll see a way of expressing that

>   The new structure of `valid(v)` is thus more restrictive for any block `v`.

Reviewer: I am failing to see the relation here as even a protocol that agrees on an invalid value would still satisfy the agreement property.

>   The new structure of `valid(v)` is thus more restrictive for any block `v`.

Author: discussed online. I'm now convince that my extra sentence (the one you proposed to remove) is not adding much, rather it risks being noise.

- consensus _validity_ is not affected since we are changing the

>   structure of the `valid(v)` predicate, on which consensus _validity_ directly depends.

Author: I see that as the reason why _agreement_ is not affected.

Validity invariant is not affected because we are changing the internal "logic" of `valid(v)"... which is, say, a "parameter" of the invariant

>   structure of the `valid(v)` predicate, on which consensus _validity_ directly depends.

Reviewer: maybe we should discuss this synchronously but I read "we are changing the "internal logic" of `valid(v)" the same as "change in the structure of the `valid(v)` predicate only makes it more restrictive", that is, the reason why _validity_ is not affected.

  In other words, for any correct process $p$, and any arbitrary block $u$,
  if $p$'s CometBFT calls `RequestProcessProposal` on $u$ at height $h$,
  then $p$'s application's acceptance or rejection **exclusively** depends on $u$ and $s_{p,h-1}$.

* Requirement 5 [`ProcessProposal`, determinism-2]:
  For any two correct processes *p* and *q*, and any arbitrary
  block $u$,
  if $p$'s (resp. $q$'s) CometBFT calls `RequestProcessProposal` on *u* at height $h$,
  then $p$'s application accepts *u* if and only if $q$'s application accepts $u$.
  Note that this requirement follows from the previous one and consensus _agreement_.

The requirements expressed above are good enough for most applications using ABCI 1.0 or 2.0.
They are simple to understand and it is relatively easy to check whether an application's
implementation of `PrepareProposal` and `ProcessProposal` fulfills them.
All applications that are able to enforce these properties do not need to reason about
the internals of the consensus implementation: they can consider it as a black box.
This is the most desirable situation in terms of modularity between CometBFT and the application.

> 

Author: Is your intention to leave this text as an annotation, or are you adding it so we can develop and integrate it in the text?

> 

Author: Answering my own question (after going through all comments) this is an annotation to develop in the text

> 

Reviewer 2: yes, exactly. I think it should be a layer of the explanation. Perhaps technical remarks. I thought annotation might be a good way of doing that. But I am not insisting this is optimal ;-)

(e.g. Proposer-Based Timestamp when expressed as an ABCI 1.0 application).

In principle, if an application's implementation of `PrepareProposal` and `ProcessProposal`
is not able to fulfill coherence and determinism requirements,
CometBFT cannot guarantee consensus _termination_ in all runs of the system.
As a result, the application designers must start considering both CometBFT and their application
as one monolithic block, in order to reason about termination.
We thus lose the modularity provided when fulfilling the ABCI 1.0 requirements.
Remember that CometBFT's consensus algorithm (Tendermint) is a well-known algorithm that
has been studied, reviewed, formally analyzed, model-checked, etc.
The combination of CometBFT and an arbitrary application as one single algorithm cannot
leverage that extensive body of research applied to the Tendermint algorithm.
This situation is risky and undesirable.

So, the questions that arises are the following.
Can we come up with a set of weaker requirements
that applications unable to fulfill the current ABCI 1.0 requirements
can still fulfill?
Is this set of weaker requirements still strong enough to guarantee consensus _termination_?

> 

Reviewer 2: ```suggestion

> Let's rephrase this a bit at the level of the consensus algorithm. Classically, Tendermint consensus 
> has been analyzed using a function `valid(v, bc(h-1))` that ensured that correct processes agree on
> the evaluation of this function because before starting height `h`, there was agreement on height 
> `h - 1`. We want to relax that by using a function `valid(v, bc(h-1), x_p)` where `x_p` is a local 
> variable at process `p`. As `x_p` may be different from `x_q` for two processes `p` and `w`, 
> using `valid` in `PrepareProposal` 
> and `ProcessProposal` may break Requirements 3 to 5.

```

Author: Makes sense. Adding the annotation text for the moment

## Solution Proposed

### Modified consensus _validity_
The specification of consensus implicitly considers `valid(v)` as an immutable function,
which is always supposed to provide the same result, when called at the same height
for the same value `v`.
This was the main reason for introducing the determinism requirements on `ProcessProposal`
in the ABCI 1.0 specification.

If we are to relax the determinism requirements on the application,
we first need to weaken the `valid(v)` predicate and therefore consensus _validity_.
In the proposed solution, correct processes MAY evaluate `valid(v)` to _true_ or _false_ at different times
for the same value `v` during height `h`.

Consensus _validity_ invariant is then modified as follows:

- _weak validity_: predicate `valid(v)` has been evaluated to _true_ at least once by one correct process

>   for the decided block `v`.

Reviewer 2: I think this write-up is a bit imprecise. It mixes "mathematical functions" (as valid) and code functions (as prepareproposal).

Author: OK, I'll see what I can do

Author: Let me know if it looks better now

Reviewer 3: Yes, I agree. If at least one correct processes says a value is valid, this requires $f+1$ processes saying it is valid.

Reviewer 4: Great, thanks!

### Eventual requirements

We now relax the relevant ABCI 1.0 requirements in the following way.

* Requirement 3b [`PrepareProposal`, `ProcessProposal`, eventual coherence]:
  There exists a round $r_s \ge 0$ of height $h$ such that,
  for any two correct processes $p$ and $q$ and any round $r_p \geq r_s$,
  if $q$'s CometBFT calls `RequestProcessProposal` on $u_p$,
  $q$'s application returns Accept in `ResponseProcessProposal`.

* The determinism-related requirements, namely requirements 4 and 5, are removed.

> Round $r_s$ is called coherence-stabilization round.

Reviewer 2: ```suggestion
We call round $r_s$ the coherence-stabilization round.
```

> 

Author: Adding annotation, will develop later

> [TODO: can we _tighten up_ eventual coherence?... i.e., make it even weaker?]
>

> > [TODO: Since we have removed determinism, can byzantine proposers cause mayhem now?]

Author: Next PI week? 😆

> > [TODO: Since we have removed determinism, can byzantine proposers cause mayhem now?]

Reviewer 3: I don't see any new treat on this aspect with the introduction of this relaxed form of validity.

I would say that this is a question we should leave open in future versions of this document.

### Modifications to the consensus Algorithm

The Tendermint algorithm as described in the [arXiv paper][arxiv],
and as implemented in CometBFT up to version `v0.38.0-rc3`,
cannot guarantee consensus _termination_ for applications

> that do not fullfil requirements 3, 4, and 5 but do fulfill requirement 3b (eventual coherence).

Reviewer 2: ```suggestion
that just fulfill requirement 3b (eventual coherence), but do not fulfill requirements 3, 4, and 5 .
```

> We need the following modifications (in terms of the algorithm as described in page 6 of the arXiv paper):

Author: Makes sense, will do. Keeping comment open for the moment

# Allowing Non-Determinism in `ProcessProposal` (a.k.a. addressing #1174)

## Context

### Consensus Properties

Byzantine Fault Tolerant (BFT) Consensus is usually specified by the following properties (2 invariants and the liveness property termination):

- _agreement_: no two correct processes decide differently.

> - _validity_: the decided block `v` satisfies the predefined predicate `valid(v)`.

Reviewer 3: This is a long discussion, but BFT consensus definitions do not necessarily include this property.

Tendermint consensus adopts a variant of the BFT consensus definition that relies on an _external_ validity check for proposed values.

Author: I took this spec from elsewhere in our own spec directory (don't remember where exactly)

- _termination_: all correct processes eventually decide.

The consensus algorithm implemented in CometBFT (Tendermint) guarantees these properties for each height.

> To be a bit more formal, valid it a mathematical function here that depends only on its inputs.

> > The arxiv paper glosses over some details here. In fact, valid is a function of the proposed block `v`

Reviewer 3: A simplified and more generic version of this is to define the predicate as `valid(v, h)`, where `h` is the current height.

If the application adopting consensus is a blockchain, as it is, to each height `h` we have associated a deterministic state `s(h)`. Plus, we have the relation `s(h) = f(s(h-1), v(h-1))`, where `v(h-1)` is the consensus output of instance `h-1`. The point is that those details might be omitted when considering the _consensus_ protocol itself, as they are "application"-specific, here the application being the blockchain.

> and the state current state of the blockchain after applying the block of height `h-1`; let's write it as
> bc(h-1). This we may write `valid(v, bc(h-1))`. Observe that this is a "mathematical function" that is
> it only depends on its input. (In contrast to functions in progamming that may use other data to 
> compute the result.

### ABCI interface

In the version of ABCI (`v0.17.0`) that existed before ABCI 1.0 and 2.0 (a.k.a. ABCI++),
the implementation of the `valid(v)` predicate was totally internal to CometBFT.
The application had no direct say on the validity of a block,

> although it could (and still can) influence it indirectly via the best-effort ABCI call `CheckTx`.

Reviewer 3: I am not sure about this last affirmation. CometBFT knows that the value proposed/decided by consensus is a block, and a block contains transactions. But we do not include, as far as I remember, in the verification of a block the verification of the transactions included in the block.

In fact, when a block is delivered to the application it is assumed that some of the transactions included in the block might be invalid, in a way that the application is expected to validate them before applying.

Author: OK, let me try to reword

With the evolution of ABCI to ABCI 1.0 and 2.0, the `valid(v)` predicate now has two components:

> - the validity checks performed directly by CometBFT on blocks

Reviewer 3: In an abstract wording, can `v` be appended to the blockchain, thus becoming the block at the current height `h`?

Author: We could, but I don't see that is needed for this write-up

  structure of the `valid(v)` predicate, on which consensus _validity_ directly depends.

However, the new structure of `valid(v)` can affect _termination_ of consensus,
as some implementations of `ProcessProposal` might reject values
that CometBFT's internal validity checks would otherwise accept.

This document focuses on how consensus _termination_ is affected
by the new structure of `valid(v)`, in particular, the different implementations of `ProcessProposal`.

### ABCI 1.0 (and 2.0) Specification

The [ABCI 1.0 specification][abci-spec] imposes a set of new requirements to the application
so that its implementation of `ProcessProposal` does not compromise _termination_ of consensus,
given the current CometBFT consensus algorithm
(called Tendermint, and described in the [arXiv paper][arxiv]).
We reproduce a slightly modified version of these requirements here.
This version of the requirements is equivalent to the ones provided in the ABCI 1.0 specification.
The modifications are needed in order to express other requirements later on in this document.

Let $p$ and $q$ be two correct processes.
Let $r_p$ be a round of consensus at height $h$ where $p$ is the proposer.

> Let $s_{p,h-1}$ be $p$'s application's state committed for height $h-1$.

Reviewer 3: A required assumption is that for every correct $p$ and $q$, $s_{p,h-1} = s_{q,h-1}$. This would allow us to remove the node indexer of the expression and render $s_{h-1}$ the same as the $bc(h-1)$ concept above proposed.

Author: I prefer to keep the same notation as in the ABCI spec. This way the reader (if they want) can compare them more easily

returns via `ResponsePrepareProposal` to CometBFT in round $r_p$, height $h$,
known as the _prepared proposal_.

These are the relevant requirements in the specification of ABCI 1.0 (and 2.0):

* Requirement 3 [`PrepareProposal`, `ProcessProposal`, coherence]: For any two correct processes $p$ and $q$
  and any round $r_p \geq 0$,
  if $q$'s CometBFT calls `RequestProcessProposal` on $u_p$,
  $q$'s application returns Accept in `ResponseProcessProposal`.

* Requirement 4 [`ProcessProposal`, determinism-1]:
  `ProcessProposal` is a (deterministic) function of the current
  state and the block being processed.
  In other words, for any correct process $p$, and any arbitrary block $u$,
  if $p$'s CometBFT calls `RequestProcessProposal` on $u$ at height $h$,
  then $p$'s application's acceptance or rejection **exclusively** depends on $u$ and $s_{p,h-1}$.

* Requirement 5 [`ProcessProposal`, determinism-2]:
  For any two correct processes *p* and *q*, and any arbitrary
  block $u$,
  if $p$'s (resp. $q$'s) CometBFT calls `RequestProcessProposal` on *u* at height $h$,
  then $p$'s application accepts *u* if and only if $q$'s application accepts $u$.

>   Note that this requirement follows from the previous one and consensus _agreement_.

Reviewer 3: Requirement 4 is a way to achieve Requirement 5, which is the one that matters.

Author: We had a discussion on this when writing the ABCI++ spec. I'm keeping it like this not to diverge from what is written there

The requirements expressed above are good enough for most applications using ABCI 1.0 or 2.0.
They are simple to understand and it is relatively easy to check whether an application's
implementation of `PrepareProposal` and `ProcessProposal` fulfills them.
All applications that are able to enforce these properties do not need to reason about
the internals of the consensus implementation: they can consider it as a black box.
This is the most desirable situation in terms of modularity between CometBFT and the application.

> `PrepareProposal` and `ProcessProposal` are understood as implemented functions. The easiest 
> (and thus the canonical) way
> to ensure the requirements is to make sure that `PrepareProposal` only proposes blocks v that

> > satisfy a (mathematical function) `valid(v, bc(h-1))`, and `ProcessProposal` just evaluates this

Reviewer 3: This is a local coherence property, namely `ProcessProposal` at correct process $p$ must accept any possible output of `PrepareProposal` at the same process $p$.

Reviewer 3: This "local coherence" can be put together with Requirement 5 and we have the properties we want.

and `ProcessProposal` just evaluates the same function.
However, `PrepareProposal` and `ProcessProposal` MAY also use other input in their
implementation, but a priori, CometBFT only guarantees consensus termination
if these implementations still ensure the requirements.

## Problem Statement

This document is dealing with the case when an application cannot guarantee
the coherence and/or determinism requirements as stated in the ABCI 1.0 specification.

An example of this is when `ProcessProposal` needs to take inputs from third-party entities
(e.g. price oracles) that are not guaranteed to provide exactly the same values to
different processes during the same height.
Another example is when `ProcessProposal` needs to read the system clock in order to perform its checks
(e.g. Proposer-Based Timestamp when expressed as an ABCI 1.0 application).

In principle, if an application's implementation of `PrepareProposal` and `ProcessProposal`
is not able to fulfill coherence and determinism requirements,
CometBFT cannot guarantee consensus _termination_ in all runs of the system.

> As a result, the application designers must start considering both CometBFT and their application

Reviewer 2: ```suggestion
As a result, the application designers a priori must start considering both CometBFT and their application
```

as one monolithic block, in order to reason about termination.
We thus lose the modularity provided when fulfilling the ABCI 1.0 requirements.
Remember that CometBFT's consensus algorithm (Tendermint) is a well-known algorithm that
has been studied, reviewed, formally analyzed, model-checked, etc.
The combination of CometBFT and an arbitrary application as one single algorithm cannot
leverage that extensive body of research applied to the Tendermint algorithm.
This situation is risky and undesirable.

So, the questions that arise are the following.
Can we come up with a set of weaker requirements
that applications unable to fulfill the current ABCI 1.0 requirements

> can still fulfill?

Reviewer 2: ```suggestion
can still fulfill?
Can we maintain modularity with these new requirements.
```

> Let's rephrase this a bit at the level of the consensus algorithm. Classically, Tendermint consensus 
> has been analyzed using a function `valid(v, bc(h-1))` that ensured that correct processes agree on
> the evaluation of this function because before starting height `h`, there was agreement on height 
> `h - 1`. We want to relax that by using a function `valid(v, bc(h-1), x_p)` where `x_p` is a local 
> variable at process `p`. As `x_p` may be different from `x_q` for two processes `p` and `q`, 
> using `valid` in `PrepareProposal` 
> and `ProcessProposal` may break Requirements 3 to 5.

## Solution Proposed

### Modified consensus _validity_
The specification of consensus implicitly considers `valid(v)` as an immutable function,
which is always supposed to provide the same result, when called at the same height

> for the same value `v`.

Reviewer 3: The name of this is _determinism_ :D

we first need to modify function $valid()$ to be of the form $valid(v, bc_{h-1}, x_p)$,
where $x_p$ is a variable local to process $p$.
As $x_p$ may be different from $x_q$ for two processes $p$ and $q$, using
$valid(v, bc_{h-1}, x_p)$ to implement in `PrepareProposal`  and `ProcessProposal`
may break Requirements 3 to 5.

Consensus _validity_ property is then modified as follows:

- _weak validity_: function $valid(v, bc_{h-1}, x_p)$ has returned _true_ at least once
  by a correct process for the decided block $v$.

> ### Eventual requirements

Author: ```suggestion
### Eventual Requirements
```

We now relax the relevant ABCI 1.0 requirements in the following way.

* Requirement 3b [`PrepareProposal`, `ProcessProposal`, eventual coherence]:

>   There exists a round $r_s \ge 0$ of height $h$ such that,

Reviewer 2: I guess the rounds are not visible to the application, so Req. 3b breaks modularity. We might cheat by introducing time intervals instead of rounds (e.g., if PrepareProposal and ProcessProposal are called within, say, 20 seconds). But perhaps for this document it would be enough to note that modularity breaks here, and it is a research question whether we can re-establish it by coming up with a different requirement.

Author: That is a good point... the spec should still make sense even if we change the underlying algorithm (e.g. for one without rounds).
I'll try to take the "time" approach you're suggesting. Bear with me

Author: OK, I went through it and got rid of $r_s$ (check the result). However, while going through the whole text I realise we're using rounds "as a notation" to specify who the proposer is. I _think_ we could also get rid of it, but it's better to leave that part as-is to not deviate from how the requirements were worded in the ABCI spec (if we come up with a different wording, or notation, every time we will add confusion)

>   for any two correct processes $p$ and $q$ and any round $r_p \geq r_s$,

Reviewer 2: Weak Validity talks about one correct processes while Req. 3b. talks about all correct processes. I guess it would make sense to highlight that?

Author: Thanks, I'll mention it

  There exists a round $r_s \ge 0$ of height $h$ such that,
  for any two correct processes $p$ and $q$ and any round $r_p \geq r_s$,

>   if $q$'s CometBFT calls `RequestProcessProposal` on $u_p$,

Reviewer 3: To be clear here: $u_p$ is produced at which round?

The definition of $u_p$ considers that it is produced at round $r_p$.

Reviewer 3: Again, if we use a "local coherence" property, we could rephrase Requirement 5 with the inclusion of this stabilization round.

Author: Good points. Check my answer to @josef-widder [above](https://github.com/cometbft/knowledge-base/pull/14/files#r1317022025). I'm gonna try to get rid of rounds

* The determinism-related requirements, namely requirements 4 and 5, are removed.

We call round $r_s$ the coherence-stabilization round.

If we think in terms of $valid(v, bc_{h-1}, x_p)$, notice that it is the application's responsibility
to ensure 3b, that is, the application designers need to prove that the $x_p$ values at correct processes

> are evolving in a way that eventually `ResponseProcessProposal` returns _Accept_ at some correct process.

Reviewer 2: ```suggestion
are evolving in a way that eventually `ResponseProcessProposal` returns _Accept_ at all correct processes.
```
If I am not mistaken...

Author: yes, actually, I'll reword to "all all correct processes that call `RequestPropocessPropsal` from that point on

> are over time evolving
> in a way that eventually ResponseProcessProposal returns ACCEPT.
> In analogy to proposer based time, `x_p` can be considered to be process `p`s local clock, and having
> clocks synchronized was used to prove acceptance of a proposal.

> [TODO: can we _tighten up_ eventual coherence?... i.e., make it even weaker?]
>
> [TODO: Since we have removed determinism, can byzantine proposers cause mayhem now?]

### Modifications to the consensus Algorithm

The Tendermint algorithm as described in the [arXiv paper][arxiv],
and as implemented in CometBFT up to version `v0.38.0-rc3`,
cannot guarantee consensus _termination_ for applications
that just fulfill requirement 3b (eventual coherence), but do not fulfill requirements 3, 4, and 5 .

We need the following modifications (in terms of the algorithm as described in page 6 of the arXiv paper):

- remove the evaluation of `valid(v)` in lines 29, 36 and 50 (i.e. replace `valid(v)` by `true`)
- modify line 23 as follows

> _\[Original\]_ &nbsp; 23: **if** $valid(v) \land (lockedRound_p = −1 \lor lockedValue_p = v)$ **then**

> _\[Modified\]_
>

> > &nbsp; 23a: $validValMatch := (validRound_p \neq -1 \land validValue_p = v)$

Reviewer 3: So, if `validRound_p == -1` then necessarily `lockedRound_p == -1`.

Considering all the values these pair of variables (valid and locked) can have during the execution, we have that locked values/rounds are subset of valid values/rounds. Namely, `locked(v, r) => valid(v, r)`.

Author: Can we use that info to simplify the expression?

Reviewer 3: So, if $validValue_p =v$ then necessarily $validRound_p \geq 0$. Namely, the first part is redundant.

Like I suggested above, I would use this additional term to define the call to $valid(x)$, namely, either  $validValue_p =v$  or we run $valid(v)$.

>

> > &nbsp; 23b: **if** $[lockedRound_p = −1 \land (validValMatch \lor valid(v))] \lor lockedValue_p=v$ **then**

Reviewer 3: Can you elaborate the rationale here?

From my understanding, we accept values that are either locked or valid (this is introduced by the change). But why?

Reviewer 3: The only difference of locked and valid is that we voted (issued a `Precommit`) for the `lockedValue`.

Author: These are good points, and make me realise we should explain the "why" in the text. Let me see what I can do

Reviewer: The idea is that a node uses the information that a value v was considered valid before (validValue == v) even if the node itself has not voted on it before (lockedRound == -1), thus avoiding invoking the _valid_ function. In PBTS terms, this would avoid considering a good proposal invalid if it is repeated in rounds bigger than the one it was first proposed.

Reviewer 3: I think I understood the formula wrong, but we should not accept a value just because it matches our valid value.

For the part that we might skip the (non-deterministic) validity check in this case, I agree. I still believe that the (deterministic) validity check must be performed at least once for a committed block.

# Allowing Non-Determinism in `ProcessProposal` (a.k.a. addressing #1174)

## Context

### Consensus Properties

Byzantine Fault Tolerant (BFT) Consensus is usually specified by the following properties.
For every height $h$:

- _agreement_: no two correct processes decide differently.
- _validity_: function $valid(v, bc_{h-1})$, when applied to the decided block, always returns _true_.
- _termination_: all correct processes eventually decide.

> The _validity_ property refers to $valid(v, bc_{h-1})$, which is defined as a (mathematical) function

Author: (lásaro: move this above the properties)

Author: done

the implementation of function $valid(v, bc_{h-1})$ was totally internal to CometBFT.
Technically, the application's part of $bc_{h-1}$ was not considered by function $valid()$.
Thus, the application had no direct say on the validity of a block,
although it could (and still can) indirectly influence the contents of blocks via the (best-effort) ABCI call `CheckTx`.

With the evolution of ABCI to ABCI 1.0 and 2.0, CometBFT's implementation of
function $valid(v, bc_{h-1})$ now has two components:

- the validity checks performed directly by CometBFT on blocks
  (block format, hashes, etc; the same as in ABCI `v0.17.0`)
- the validity checks that now the application can perform as part of `ProcessProposal`;
  i.e., `ProcessProposal` is now part of $valid(v, bc_{h-1})$

With the new structure of the implementation of function $valid(v, bc_{h-1})$:

- consensus _agreement_ is not affected and all processes are still required to agree on the same value
- the consensus _validity_ property is not affected since we are changing the
  internals of function $valid(v, bc_{h-1})$; consensus _validity_ just requires this function to be true

> However, the new structure of the implementation of function $valid(v, bc_{h-1})$

Reviewer: Termination can only be achieved (while satisfying the other properties) if valid proposals are made. This has not changed with the new specification. What has changed, as you have pointed elsewhere, is that valid is possibly more restrictive now. Maybe this should be clarified here.

Author: Added sentence to clarify

may affect _termination_ of consensus, as some implementations of `ProcessProposal` might reject values
that CometBFT's internal validity checks would otherwise accept.

This document focuses on how consensus _termination_ is affected
by the new structure of function $valid(v, bc_{h-1})$,
in particular, the different implementations of `ProcessProposal`.

### ABCI 1.0 (and 2.0) Specification

The [ABCI 1.0 specification][abci-spec] imposes a set of new requirements on the application
so that its implementation of `PrepareProposal` and `ProcessProposal` does not compromise _termination_ of consensus,
given the current CometBFT consensus algorithm
(called Tendermint, and described in the [arXiv paper][arxiv]).
In contrast to $valid(v, bc_{h-1})$, which is defined as a mathematical function used for consensus's formal specification,

> `PrepareProposal` and `ProcessProposal` are understood as implemented functions in CometBFT.

Reviewer: I believe this has to be reviewed, in the light of the more recent changes. valid now is split into two components. Also, nowhere so far it has been said that ProcessProposal does not have to be a pure function.

Author: What would you suggest as text?

Author: "implemented function" means function in its software (rather than mathematical) meaning. I'll try to reword

known as the _raw proposal_.
Let $u_p$ be the (possibly modified) block that $p$'s application
returns via `ResponsePrepareProposal` to CometBFT in round $r_p$, height $h$,
known as the _prepared proposal_.

* Requirement 3 [`PrepareProposal`, `ProcessProposal`, coherence]: For any two correct processes $p$ and $q$
  if $q$'s CometBFT calls `RequestProcessProposal` on $u_p$,
  $q$'s application returns _Accept_ in `ResponseProcessProposal`.

* Requirement 4 [`ProcessProposal`, determinism-1]:
  `ProcessProposal` is a (deterministic) function of the current
  state and the block being processed.
  In other words, for any correct process $p$, and any arbitrary block $u$,
  if $p$'s CometBFT calls `RequestProcessProposal` on $u$ at height $h$,
  then $p$'s application's acceptance or rejection **exclusively** depends on $u$ and $s_{p,h-1}$.

* Requirement 5 [`ProcessProposal`, determinism-2]:
  For any two correct processes *p* and *q*, and any arbitrary
  block $u$,
  if CometBFT instances at $p$ and $q$ call `RequestProcessProposal` on $u$ at height $h$,
  then $p$'s application accepts $u$ if and only if $q$'s application accepts $u$.

>   Note that this requirement follows from the previous one and consensus _agreement_.

Author: Yes

The combination of CometBFT and an arbitrary application as one single algorithm cannot
leverage that extensive body of research applied to the Tendermint algorithm.
This situation is risky and undesirable.

So, the questions that arise are the following.
Can we come up with a set of weaker requirements
that applications unable to fulfill the current ABCI 1.0 requirements
can still fulfill?
Can we maintain modularity with these new requirements?
Is this set of weaker requirements still strong enough to guarantee consensus _termination_?

## Solution Proposed

### Modified consensus _validity_

Function $valid(v, bc_{h-1})$, as explained above, exclusively depends on its inputs
(a block, and the blockchain state at the previous height).
So it is always supposed to provide the same result when called at the same height for the same inputs,
no matter at which process.
This was the main reason for introducing the determinism requirements on `ProcessProposal`
in the ABCI 1.0 specification.

If we are to relax the determinism requirements on the application,
we first need to modify function $valid()$ to be of the form $valid(v, bc_{h-1}, x_p)$,

> where $x_p$ is a variable local to process $p$.

Author: Yes. Let me add this sentence to the text

As $x_p$ may be different from $x_q$ for two processes $p$ and $q$, using
$valid(v, bc_{h-1}, x_p)$ to implement in `PrepareProposal`  and `ProcessProposal`
may break Requirements 3 to 5.

Consensus _validity_ property is then modified as follows:

- _weak validity_: function $valid(v, bc_{h-1}, x_p)$ has returned _true_ at least once
  by a correct process for the decided block $v$.

### Eventual Requirements

We now relax the relevant ABCI 1.0 requirements in the following way.

* Requirement 3b [`PrepareProposal`, `ProcessProposal`, eventual coherence]:
  There exists a time $ts_{h}$ for every height $h$ such that,
  for any two correct processes $p$ and $q$ and any round $r_p$ in height $h$ starting after $ts_{h}$,
  if $q$'s CometBFT calls `RequestProcessProposal` on $u_p$,
  $q$'s application returns _Accept_ in `ResponseProcessProposal`.

* The determinism-related requirements, namely requirements 4 and 5, are removed.

> We call $ts_{h}$ the coherence-stabilization time for height $h$.

Reviewer 3: Post-comment: the problem addressed here, namely how a recovery node running at a different time from the quorum of the up-to-dated nodes should proceed, does not depend on whether we adopt physical or logical (e.g. round) time in this definition.

All in all, the provided definition is good enough for establishing that at least a quorum o correct nodes converge on the acceptance of a block produced by a correct instance of the application.

> We call $ts_{h}$ the coherence-stabilization time for height $h$.

Author: As discussed synchronously, I'll add a sentence here with a link to cometbft/cometbft#1035 saying that it is related, and, for the time being, CometBFT signer is good enough in the implementation

If we think in terms of $valid(v, bc_{h-1}, x_p)$, notice that it is the application's responsibility
to ensure 3b, that is, the application designers need to prove that the $x_p$ values at correct processes
are evolving in a way that eventually `ResponseProcessProposal` returns _Accept_ at all correct processes
that call `RequestPropocessPropsal` for height $h$ after $ts_{h}$.
For instance, in Proposer-Based Timestamp, $x_p$ can be considered to be process $p$'s local clock,
and having clocks synchronized is the mechanism ensuring eventual acceptance of a proposal.
Finally, it is worth noting that _weak validity_ requires just one correct processes while requirement 3b
refers to all correct processes.

> > [TODO: can we _tighten up_ eventual coherence?... i.e., make it even weaker?]

Reviewer 3: I can hardly imagine a weaker form for it. We are already defining a valid value as a value that has been found valid by a single correct process. In the implementation we are stronger, as we require 2f+1, namely f+1 correct.

Reviewer: The only way in which I can think is to make the common acceptance happen infinitely often rather than in every round after $ts_h$.

>
> [TODO: Since we have removed determinism, can byzantine proposers cause mayhem now?]

### Modifications to the consensus Algorithm

The Tendermint algorithm as described in the [arXiv paper][arxiv],
and as implemented in CometBFT up to version `v0.38.0-rc3`,
cannot guarantee consensus _termination_ for applications
that just fulfill requirement 3b (eventual coherence), but do not fulfill requirements 3, 4, and 5.

We need the following modifications (in terms of the algorithm as described in page 6 of the arXiv paper):

- remove the evaluation of `valid(v)` in lines 29, 36 and 50 (i.e. replace `valid(v)` by `true`)
- modify line 23 as follows

> > _\[Original\]_ &nbsp; 23: **if** $valid(v) \land (lockedRound_p = −1 \lor lockedValue_p = v)$ **then**

Reviewer 3: Ok, lets go through the update process here:

The original version is: $valid(v) \land (lockedRound_p = −1 \lor lockedValue_p = v)$.

This is the same as: $(valid(v) \land lockedRound_p = −1) \lor (valid(v) \land \lor lockedValue_p = v)$.

The first change is that we don't need to check the validity of a locked value. From this we get: $(valid(v) \land lockedRound_p = −1) \lor lockedValue_p = v$.

The second change is to avoid checking validity for a valid value, so: $((validValue_p = v \lor valid(v)) \land lockedRound_p = −1) \lor lockedValue_p = v$.

An alternative is to replace $valid(v)$ by a new version, say $valid*(v)$, so that we get back to a simpler representation: $(valid*(v) \land lockedRound_p = −1) \lor lockedValue_p = v$, where we define $valid*(v) = validValue = v \lor valid(v)$.

Reviewer 3: Post-comment: the formulation on the document is **right**. This is a proposal for making it simpler.

> _\[Modified\]_
>
> &nbsp; 23a: $validValMatch := (validRound_p \neq -1 \land validValue_p = v)$
>
> &nbsp; 23b: **if** $[lockedRound_p = −1 \land (validValMatch \lor valid(v))] \lor lockedValue_p=v$ **then**

- The occurrences of `valid(v)` that we have removed were in a way redundant,

>   so removing them does not affect the ability of the algorithm to fulfill consensus properties.

Reviewer 3: Here you are considering the _weak validity_ property, not the original _validity_ property.

For the latter, if we commit (or act in a way to enable committing) $v$, then we necessarily have to locally verify $v$ by running $valid(v)$.

For the former, the introduced variation, we are removing this method when we have 2f+1 Prevotes for $v$, either explicitly in the same upon clause (lines 29 and 36), or implicitly (line 50).

Author: Good point. Reworded the sentence

as part of issues [#1171][1171], and [#1230][1230].

## Conclusion

This document has explored the possibility of relaxing the coherence and determinism properties
of the ABCI 1.0 (and 2.0) specification affecting `PrepareProposal` and `ProcessProposal`
for a class of applications that cannot guarantee them.

We first weakened the _validity_ property of the consensus specification
in a way that keeps the overall consensus specification strong enough to be relevant.
We then proposed a weaker coherence property for ABCI 1.0 (and 2.0) that can replace the original
coherence and determinism properties related to `PrepareProposal` and `ProcessProposal`.
The new property is useful for applications that cannot fulfill the original properties
but can fulfill the new one.
Finally, we explained how to modify the Tendermint consensus algorithm to guarantee
the consensus _termination_ property for applications that fulfill the new property.

In this document, we have not tackled the problem of applications
that cannot fulfill coherence and determinism properties that refer to vote extensions in the ABCI 2.0 specification.

> We leave this as future work.

Reviewer 3: I would at least mention that we need to find a way to properly address those changes in the case of a recovering process.

The case to be considered is when the application accepted a value during regular operation and may reject the same value during recovery, or vice-versa. This possibility has the potential of producing conflicting Prevote messages (accepting or rejecting) a value during recovery.

Reviewer 3: Note that the same safe guard discussed in https://github.com/cometbft/cometbft/issues/1035 should prevent nodes from equivocating in recovery mode.

Author: Yes, added a section on this

(e.g. price oracles) that are not guaranteed to provide exactly the same values to
different processes during the same height.
Another example is when `ProcessProposal` needs to read the system clock in order to perform its checks
(e.g. Proposer-Based Timestamp when expressed as an ABCI 1.0 application).

In principle, if an application's implementation of `PrepareProposal` and `ProcessProposal`
is not able to fulfill coherence and determinism requirements,
CometBFT cannot guarantee consensus _termination_ in all runs of the system.
As a result, the application designers a priori must start considering both CometBFT and their application
as one monolithic block, in order to reason about termination.
We thus lose the modularity provided when fulfilling the ABCI 1.0 requirements.
Remember that CometBFT's consensus algorithm (Tendermint) is a well-known algorithm that
has been studied, reviewed, formally analyzed, model-checked, etc.
The combination of CometBFT and an arbitrary application as one single algorithm cannot
leverage that extensive body of research applied to the Tendermint algorithm.

> This situation is risky and undesirable.

Author: I'd like to keep it, as I think it is important to stress that reasoning about termination of a distributed algorithms isn't something that can be done lightly. I'll accept you suggestion (with the "Warning") but will revert it to normal text
